Move Vector Embeddings Guide from HANA to Databases (#2507)
cap-js/docs: guides/databases/vector-embeddings.md

```cds
extend Incidents with {
  @cds.api.ignore
  embedding : Vector(768) = vector_embedding( 
    'Title: ' || title || ', Summary: ' || summary, 
    'DOCUMENT', 'SAP_GXY.20250407'
  ) stored; 
}
```

> [!warning] Java only and <Beta/>
> The `vector_embedding` function is currently in beta and only supported by the CAP Java runtime.

::: info Local Testing with H2 and SQLite
On H2 and SQLite the `CQL.vectorEmbedding` function is emulated to support local testing.
:::

### Generate Embeddings Programmatically

Alternatively, you can compute vector embeddings in your application layer using the [SAP Cloud SDK for AI](https://sap.github.io/ai-sdk/) to call SAP AI Core services for generating embeddings.

:::details Example using SAP Cloud SDK for AI
```Java
var aiClient = OpenAiClient.forModel(OpenAiModel.TEXT_EMBEDDING_3_SMALL);
var response = aiClient.embedding(
   new OpenAiEmbeddingRequest(List.of(book.getDescription())));
book.setEmbedding(CdsVector.of(response.getEmbeddingVectors().get(0)));
```
:::

> 

Author: ```suggestion

:::tip Use SAP Cloud SDK for AI
Use the [SAP Cloud SDK for AI](https://sap.github.io/ai-sdk/) for unified access to embedding models and large language models (LLMs) from [SAP AI Core](https://help.sap.com/docs/sap-ai-core/sap-ai-core-service-guide/what-is-sap-ai-core).
:::

Learn more about the [SAP Cloud SDK for AI (Java)](https://sap.github.io/ai-sdk/docs/java/getting-started) or the [SAP Cloud SDK for AI (JavaScript)](https://sap.github.io/ai-sdk/docs/js/getting-started) {.learn-more}

```

// Find Incidents related to user question ordered by relevance
Select.from(INCIDENTS)
   .columns(i -> similarity.times(100).as("relevance"), 
            i -> i.ID(), i -> i.title(), i -> i.summary(), i -> i.date())
   .where(i -> similarity.gt(0.75f))
   .orderBy(i -> i.get("relevance").desc());
```
```

```js [Node.js]
const response = await new AzureOpenAiEmbeddingClient(
 'text-embedding-3-small'
).run({
 input: 'Any incidents with solar inverters this month? How were they resolved?'
});

const questionEmbedding = response.getEmbedding();
let similarIncidents = await SELECT.from('Incidents')
  .where`cosine_similarity(embedding, to_real_vector(${questionEmbedding})) > 0.75`;
```
:::

:::tip Evolve embeddings with your model
Store embeddings when you create or update your data. Regenerate embeddings if you change your embedding model.
:::

:::tip Use SAP Cloud SDK for AI
Use the [SAP Cloud SDK for AI](https://sap.github.io/ai-sdk/) for unified access to embedding models and large language models (LLMs) from [SAP AI Core](https://help.sap.com/docs/sap-ai-core/sap-ai-core-service-guide/what-is-sap-ai-core).
:::

> Learn more about the [SAP Cloud SDK for AI (Java)](https://sap.github.io/ai-sdk/docs/java/getting-started) or the [SAP Cloud SDK for AI (JavaScript)](https://sap.github.io/ai-sdk/docs/js/getting-started) {.learn-more}

Author: ```suggestion
```

var query = CQL.val(
  "Any incidents with solar inverters this month? How were they resolved?");
var embedding = CQL.vectorEmbedding(query, TextType.QUERY, "SAP_GXY.20250407");

// Compute similarity between user question and incident embeddings
var similarity = CQL.cosineSimilarity(CQL.get(Incidents.EMBEDDING), embedding);

// Find Incidents related to user question ordered by relevance
Select.from(INCIDENTS)
   .columns(i -> similarity.times(100).as("relevance"), 
            i -> i.ID(), i -> i.title(), i -> i.summary(), i -> i.date())
   .where(i -> similarity.gt(0.75f))
   .orderBy(i -> i.get("relevance").desc());
```
```

```js [Node.js]
const response = await new AzureOpenAiEmbeddingClient(
 'text-embedding-3-small'
).run({
 input: 'Any incidents with solar inverters this month? How were they resolved?'
});

const questionEmbedding = response.getEmbedding();
let similarIncidents = await SELECT.from('Incidents')
  .where`cosine_similarity(embedding, to_real_vector(${questionEmbedding})) > 0.75`;
```
:::

:::tip Evolve embeddings with your model
Store embeddings when you create or update your data. Regenerate embeddings if you change your embedding model.

> :::

Author: ```suggestion
```

## Choose an Embedding Model

Choose an embedding model that fits your use case and data (for example english or multilingual text). The model determines the number of dimensions of the resulting output vector. Check the documentation of the respective embedding model for details.

Use the [SAP Generative AI Hub](https://www.sap.com/products/artificial-intelligence/generative-ai-hub.html) for unified consumption of embedding models and LLMs across different vendors and open source models. Check for available models on the [SAP AI Launchpad](https://help.sap.com/docs/ai-launchpad/sap-ai-launchpad-user-guide/models-and-scenarios-in-generative-ai-hub-fef463b24bff4f44a33e98bb1e4f3148#models).

## Add Embeddings to Your CDS Model
Use the built-in CDL [Vector](../cds/types) type in your CDS model to store embeddings. Set the vector dimensions to match the embedding model (for example, 768 for *SAP_GXY.20250407*).

```cds
extend Incidents with {

>   embedding : cds.Vector(768);

Author: ```suggestion
  embedding : Vector(768);
```

[Learn more about Vector Embeddings in CAP Java](../../java/cds-data#vector-embeddings) {.learn-more}

### Generate Embeddings Programmatically

Alternatively, you can compute vector embeddings in your application layer using the [SAP Cloud SDK for AI](https://sap.github.io/ai-sdk/) to call SAP AI Core services for generating embeddings.

:::details Example using SAP Cloud SDK for AI
```Java
var aiClient = OpenAiClient.forModel(OpenAiModel.TEXT_EMBEDDING_3_SMALL);
var response = aiClient.embedding(
   new OpenAiEmbeddingRequest(List.of(book.getDescription())));
book.setEmbedding(CdsVector.of(response.getEmbeddingVectors().get(0)));
```
:::

:::tip Use SAP Cloud SDK for AI
Use the [SAP Cloud SDK for AI](https://sap.github.io/ai-sdk/) for unified access to embedding models and large language models (LLMs) from [SAP AI Core](https://help.sap.com/docs/sap-ai-core/sap-ai-core-service-guide/what-is-sap-ai-core).
:::

Learn more about the [SAP Cloud SDK for AI (Java)](https://sap.github.io/ai-sdk/docs/java/getting-started) or the [SAP Cloud SDK for AI (JavaScript)](https://sap.github.io/ai-sdk/docs/js/getting-started) {.learn-more}

## Query for Similarity

> At runtime, use vector functions to search for similar items. For example, find incidents that are relevant to a user question:

Author: ```suggestion
At runtime, use vector functions to search for similar items. In a Retrieval-Augmented Generation (RAG) scenario, for example to enhance the context of a user query for the LLM to answer the query, compute the vector embedding of the user query and use `CQL.cosineSimilarity` to find incidents that are relevant to a user question and can be given to the LLM to answer the question.
```

::: code-group
```Java [Java]
// Compute embedding for user question
var query = CQL.val(
  "Any incidents with solar inverters this month? How were they resolved?");
var embedding = CQL.vectorEmbedding(query, TextType.QUERY, "SAP_GXY.20250407");

// Compute similarity between user question and incident embeddings
var similarity = CQL.cosineSimilarity(CQL.get(Incidents.EMBEDDING), embedding);

// Find Incidents related to user question ordered by relevance
Select.from(INCIDENTS)
   .columns(i -> similarity.times(100).as("relevance"), 
            i -> i.ID(), i -> i.title(), i -> i.summary(), i -> i.date())

>    .where(i -> similarity.gt(0.75f))

Author: ```suggestion
   .where(i -> similarity.gt(0.75))
```

---
label: Vector Embeddings
---
# Vector Embeddings

Vector embeddings let you add semantic search, recommendations, and generative AI features to your CAP application. Embeddings are numeric arrays that represent the meaning of unstructured data (text, images, etc.), making it possible to compare and search for items that are semantically related to each other or a user query.

## Choose an Embedding Model

Choose an embedding model that fits your use case and data (for example english or multilingual text). The model determines the number of dimensions of the resulting output vector. Check the documentation of the respective embedding model for details.

> Use the [SAP Generative AI Hub](https://community.sap.com/t5/technology-blogs-by-sap/how-sap-s-generative-ai-hub-facilitates-embedded-trustworthy-and-reliable/ba-p/13596153) for unified consumption of embedding models and LLMs across different vendors and open source models. Check for available models on the [SAP AI Launchpad](https://help.sap.com/docs/ai-launchpad/sap-ai-launchpad-user-guide/models-and-scenarios-in-generative-ai-hub-fef463b24bff4f44a33e98bb1e4f3148#models).

Author: ```suggestion
Use the [SAP Generative AI Hub](https://www.sap.com/products/artificial-intelligence/generative-ai-hub.html) for unified consumption of embedding models and LLMs across different vendors and open source models. Check for available models on the [SAP AI Launchpad](https://help.sap.com/docs/ai-launchpad/sap-ai-launchpad-user-guide/models-and-scenarios-in-generative-ai-hub-fef463b24bff4f44a33e98bb1e4f3148#models).
```

---
label: Vector Embeddings
---
# Vector Embeddings

> Vector embeddings convert unstructured content (text, images, etc.) into numeric vectors that encode semantic meaning. Comparing these vectors enables semantic search, recommendations, and enhanced generative AI features in your CAP application, for example retrieving related records, ranking results by relevance, or augmenting prompts for LLMs.

Reviewer: ```suggestion
Vector embeddings convert unstructured content (text, images, etc.) into numeric vectors that encode semantics (meaning). Comparing these vectors enables semantic search, recommendations, and enhanced generative AI features in your CAP application, for example retrieving related records, ranking results by relevance, or augmenting prompts for LLMs.
```

"semantics" - "meaning"
"semantic" - "connected with meaning"
"semantic search" - "search by meaning"
"semantic meaning" - ?

## Choose an Embedding Model

Choose an embedding model that fits your use case and data (for example english or multilingual text). The model determines the number of dimensions of the resulting output vector. Check the documentation of the respective embedding model for details.

Use the [SAP Generative AI Hub](https://www.sap.com/products/artificial-intelligence/generative-ai-hub.html) for unified consumption of embedding models and LLMs across different vendors and open source models. Check for available models on the [SAP AI Launchpad](https://help.sap.com/docs/ai-launchpad/sap-ai-launchpad-user-guide/models-and-scenarios-in-generative-ai-hub-fef463b24bff4f44a33e98bb1e4f3148#models).

## Add Embeddings to Your CDS Model

> Use the `cds.Vector` type in your CDS model to store embeddings on SAP HANA Cloud. Set the dimension to match your embedding model (for example, 768 embedding dimensions for *SAP_GXY.20250407*).

Author: ```suggestion
Use the built-in CDL [Vector](../cds/types) type in your CDS model to store embeddings. Set the vector dimensions to match the embedding model (for example, 768 for *SAP_GXY.20250407*).
```

```cds
extend Incidents with {

>   embedding : cds.Vector(768);

Reviewer: I think it would be very useful to annotate the model to be used.

}
```

## Generate Embeddings
Use an embedding model to convert your data (for example, incident summaries) into vectors.

> 

Author: ```suggestion

### Generate Embeddings on the Database

```

To generate vector embeddings on write in SAP HANA, you can use the [vector_embedding](https://help.sap.com/docs/hana-cloud-database/sap-hana-cloud-sap-hana-database-sql-reference-guide/vector-embedding-function-vector) function as calculated element [on-write](../../cds/cdl#on-write) with embedding models from [SAP HANA NLP](https://help.sap.com/docs/hana-cloud-database/sap-hana-cloud-sap-hana-database-vector-engine-guide/creating-text-embeddings-with-nlp-51eb170d038d4099a9bbb85c08fda888) or a configured remote source from SAP AI Core:

```cds
extend Incidents with {
  embedding : cds.Vector(768) = vector_embedding(

>        summary, 'DOCUMENT', 'SAP_GXY.20250407') stored;

Reviewer: @cds.api.ignore

}
```

> 

Author: ```suggestion

:::info beta
The `vector_embedding` function is currently in beta and only supported by the CAP Java runtime.
:::

```

Reviewer: I'd rather add subsection 

### Generate Embeddings on the Database <beta />

(On SAP HANA it's actually not beta as we can't prevent anyone from using the vector_embedding function). It's beta on H2 and SQLite. On Postgres not supported.

But if we follow-up with the ID to introduce symbolic names for the models it's beta also on HANA.

### Generate Embeddings Programatically

> Alternatively, you can compute vector embeddings in your application layer using the [SAP Cloud SDK for AI](https://sap.github.io/ai-sdk/) to call SAP AI Core services for generating embeddings.

Author: ```suggestion
### Generate Embeddings Programmatically

Alternatively, you can compute vector embeddings in your application layer using the [SAP Cloud SDK for AI](https://sap.github.io/ai-sdk/) to call SAP AI Core services for generating embeddings.
```

:::details Example using SAP Cloud SDK for AI
```Java
var aiClient = OpenAiClient.forModel(OpenAiModel.TEXT_EMBEDDING_3_SMALL);
var response = aiClient.embedding(
   new OpenAiEmbeddingRequest(List.of(book.getDescription())));
book.setEmbedding(CdsVector.of(response.getEmbeddingVectors().get(0)));
```
:::

## Query for Similarity

> At runtime, use SAP HANA's built-in vector functions to search for similar items. For example, find incidents that are relevant to a user question:

Reviewer: ```suggestion
At runtime, use vector functions to search for similar items. For example, find incidents that are relevant to a user question:
```

::: code-group
```Java [Java]
// Compute embedding for user question
var query = CQL.val(
  "Any incidents with solar inverters this month? How were they resolved?");

> var embedding = CQL.vectorEmbedding(query, QUERY, "SAP_GXY.20250407");

Author: ```suggestion
var embedding = CQL.vectorEmbedding(query, TextType.QUERY, "SAP_GXY.20250407");
```

// Compute similarity between user question and incident embeddings
var similarity = CQL.cosineSimilarity(CQL.get(Incidents.EMBEDDING), embedding);

// Find Incidents related to user question ordered by relevance
Select.from(INCIDENTS)
   .columns(i -> similarity.times(100).as("relevance"), 
            i -> i.ID(), i -> i.title(), i -> i.summary(), i -> i.date())

>    .where(i -> similarity.gt(minSimilarity))

Author: ```suggestion
   .where(i -> similarity.gt(0.75f))
```

>    .orderBy(i -> i.get("relevance").desc());

Reviewer: or

```suggestion
   .orderBy(i -> i.get("relevance").desc())
   .limit(5);
```

Author: This would also return results with very low similarity causing noise

```
```

```js [Node.js]
const response = await new AzureOpenAiEmbeddingClient(
 'text-embedding-3-small'
).run({
 input: 'Any incidents with solar inverters this month? How were they resolved?'
});

const questionEmbedding = response.getEmbedding();
let similarIncidents = await SELECT.from('Incidents')

>   .where`cosine_similarity(embedding, to_real_vector(${questionEmbedding})) > 0.9`;

Reviewer: >  0.9

I could imagine that a similarity of 0.9 is hard to achieve

Use the same value in the Java and Node.js exmaples.

Consider to make the examples equivalent. Also consider to have a bilingual code box

>   .where`cosine_similarity(embedding, to_real_vector(${questionEmbedding})) > 0.9`;

Author: ```suggestion
  .where`cosine_similarity(embedding, to_real_vector(${questionEmbedding})) > 0.75`;
```

```
:::

:::tip Evolve embeddings with your model
Store embeddings when you create or update your data. Regenerate embeddings if you change your embedding model.
:::

> 

Reviewer: This applies to programmatic embedding only. Add a similar tip-box to for the DB-generated embedding and promote the advantage of  stored calculated elements / GENERATED ALWAYS

## Choose an Embedding Model

Choose an embedding model that fits your use case and data (for example english or multilingual text). The model determines the number of dimensions of the resulting output vector. Check the documentation of the respective embedding model for details.

Use the [SAP Generative AI Hub](https://www.sap.com/products/artificial-intelligence/generative-ai-hub.html) for unified consumption of embedding models and LLMs across different vendors and open source models. Check for available models on the [SAP AI Launchpad](https://help.sap.com/docs/ai-launchpad/sap-ai-launchpad-user-guide/models-and-scenarios-in-generative-ai-hub-fef463b24bff4f44a33e98bb1e4f3148#models).

## Add Embeddings to Your CDS Model
Use the built-in CDL [Vector](../cds/types) type in your CDS model to store embeddings. Set the vector dimensions to match the embedding model (for example, 768 for *SAP_GXY.20250407*).

```cds
extend Incidents with {
  embedding : cds.Vector(768);
}
```

## Generate Embeddings

> Use an embedding model to convert your data (for example, incident summaries) into vectors.

Author: ```suggestion
Use an embedding model to convert your data (for example, incident titles and summaries) into vectors.

:::tip Evolve embeddings with your model
Store embeddings when you create or update your data. Regenerate embeddings if you change your embedding model.
:::
```

### Generate Embeddings on the Database

To generate vector embeddings on write in SAP HANA, you can use the [vector_embedding](https://help.sap.com/docs/hana-cloud-database/sap-hana-cloud-sap-hana-database-sql-reference-guide/vector-embedding-function-vector) function as calculated element [on-write](../../cds/cdl#on-write) with embedding models from [SAP HANA NLP](https://help.sap.com/docs/hana-cloud-database/sap-hana-cloud-sap-hana-database-vector-engine-guide/creating-text-embeddings-with-nlp-51eb170d038d4099a9bbb85c08fda888) or a configured remote source from SAP AI Core:

```cds
extend Incidents with {
  embedding : cds.Vector(768) = vector_embedding(

>        summary, 'DOCUMENT', 'SAP_GXY.20250407') stored;

Author: ```suggestion
  @cds.api.ignore
  embedding : Vector(768) = vector_embedding( 
    'Title: ' || title || ', Summary: ' || summary, 
    'DOCUMENT', 'SAP_GXY.20250407'
  ) stored; 
```

}
```

:::info beta
The `vector_embedding` function is currently in beta and only supported by the CAP Java runtime.

> :::

Author: ```suggestion
> [!warning] Java only and <Beta/>
> The `vector_embedding` function is currently in beta and only supported by the CAP Java runtime.
```

> 

Author: ```suggestion

::: info Local Testing with H2 and SQLite
On H2 and SQLite the `CQL.vectorEmbedding` function is emulated to support local testing.
:::

```
